Allow clicking buttons on a touch screen

TheZoq2 opened the pull request:
My main desired usecase of xfdashboard is as an app launcher and window switcher when my 2-in-1 is in tablet mode. However, tocuhes didn't seem to register as clicks until I made this very simple change.

I assume that there is some constant that would be better to use instead of `1` but I haven't looked into it.

Scrolling in menus also doesn't work on the touch screen which may or may not be related to this. Other than that, everything seems to work.

gmc-holle commented:
I believe that xfdashboard is not working (well) on touchscreens as I do not own such a device and it is not tested therefore. So thanks for testing it :)

XfdashboardClickAction is simply a copy of the original ClutterClickAction with just one function returning a different boolean value. The reason for this "copy" is that ClutterClickAction does not work well on overlapping actors on the same z-coordinate.

But I do not know if chaning the event structure is the right way. Your patch implements something I would call "touchscreen-emulates-pointer" by setting a button pressed (1 for left button) at a short touch event.

I asking myself: Would it be better to make xfdashboard touchsceen-aware? That means if a "clicked" signal is emitted the button pressed is 1 for left button and 0 for touchscreen. So all calls to xfdashboard_click_action_get_button() must check against this two values, e.g.

```
diff --git a/libxfdashboard/live-window.c b/libxfdashboard/live-window.c
index 9e87853..457d873 100644
--- a/libxfdashboard/live-window.c
+++ b/libxfdashboard/live-window.c
@@ -357,8 +357,11 @@ static void _xfdashboard_live_window_on_clicked(XfdashboardLiveWindow *self,
        priv=self->priv;
        action=XFDASHBOARD_CLICK_ACTION(inUserData);
 
-       /* Only emit any of these signals if click was perform with left button */
-       if(xfdashboard_click_action_get_button(action)!=XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON) return;
+       /* Only emit any of these signals if click was perform with left button 
+        * or is a short touchscreen touch event.
+        */
+       if(xfdashboard_click_action_get_button(action)!=XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON &&
+               xfdashboard_click_action_get_button(action)!=0) return;
 
        /* Check if click happened in "close button" */
        if(clutter_actor_is_visible(priv->actorClose))
```

On the other side your way is much simpler as it "says" that a short touchscreen event is a left-button click.

What do you think?

TheZoq2 commented:
Yep, that sounds like a better solution. I tried implementing it and it works for windows, but not  the other buttons in the program. I assume that's because live-windoww.c only handles windows and I should make the change in other places as well?

gmc-holle commented:
This was just an example and affects only windows. There are also workspaces, button etc.

A simple grep in folder libxfdashboard reveals these files needed to be modified:

```
$ grep xfdashboard_click_action_get_button libxfdashboard/*.c | grep -v click-action.c
libxfdashboard/button.c:	if(xfdashboard_click_action_get_button(inAction)==XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON)
libxfdashboard/live-window.c:	if(xfdashboard_click_action_get_button(action)!=XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON) return;
libxfdashboard/live-workspace.c:	if(xfdashboard_click_action_get_button(action)==XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON)
libxfdashboard/popup-menu-item-button.c:	if(xfdashboard_click_action_get_button(inAction)==XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON)
libxfdashboard/quicklaunch.c:	if(xfdashboard_click_action_get_button(action)==XFDASHBOARD_CLICK_ACTION_RIGHT_BUTTON)
libxfdashboard/search-result-container.c:	if(xfdashboard_click_action_get_button(inAction)==XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON)
```

TheZoq2 commented:
Yep, I managed to find it, thanks. Making these changes fixed everything except for tapping on workspaces. Did I miss anything.

Also, this feels like a place where a utility function could be used since this code is duplicated everywhere. However, im not sure where I would put that.

gmc-holle commented on libxfdashboard/live-workspace.c:
```diff
@@ -345,8 +345,11 @@ static void _xfdashboard_live_workspace_on_clicked(XfdashboardLiveWorkspace *sel
+	/* Only emit any of these signals if click was perform with left button 
+	 * or is a short touchscreen touch event.
+	 */
+	if(xfdashboard_click_action_get_button(action)!=XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON &&
```
This is not correct replaced. The original code checks for left-button but the replaced code checks for non-left-button.

It should be:

```
if(xfdashboard_click_action_get_button(action)==XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON ||
    xfdashboard_click_action_get_button(action)==0)
```

gmc-holle commented on libxfdashboard/popup-menu-item-button.c:
```diff
@@ -63,8 +63,11 @@ static void _xfdashboard_popup_menu_item_button_clicked(XfdashboardClickAction *
+	 * or is a short touchscreen touch event.
+	 */
+	if(xfdashboard_click_action_get_button(inAction)!=XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON &&
+			xfdashboard_click_action_get_button(inAction)!=0) return;
```
Same as above. Check for left-button presses (or touch event) and not for non-left-button events.

gmc-holle commented on libxfdashboard/quicklaunch.c:
```diff
@@ -671,8 +671,11 @@ static void _xfdashboard_quicklaunch_on_favourite_popup_menu(XfdashboardQuicklau
+	 * or is a short touchscreen touch event.
+	 */
+	if(xfdashboard_click_action_get_button(action)!=XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON &&
+			xfdashboard_click_action_get_button(action)!=0) return;
```
Same as above but this time check for right button. I don't know how to catch right-button pressed on touchscreen as long-press touch event are most likely be used for drag-and-drop actions on touchscreen, aren't they?

gmc-holle commented on libxfdashboard/search-result-container.c:
```diff
@@ -393,8 +393,11 @@ static void _xfdashboard_search_result_container_on_result_item_actor_clicked(Xf
+	 * or is a short touchscreen touch event.
+	 */
+	if(xfdashboard_click_action_get_button(inAction)!=XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON &&
+			xfdashboard_click_action_get_button(inAction)!=0) return;
```
Same as above.

gmc-holle commented:
Maybe a utility function will help. My idea would be to add the utility function to XfdashboardClickAction:

```
gboolean xfdashboard_click_action_is_left_button_or_touch(XfdashboardClickAction *self)
{
  XfdashboardClickActionPrivate *priv;

  g_return_val_if_fail(XFDASHBOARD_IS_CLICK_ACTION(self), FALSE);

  priv=self->priv;

  if(priv->isPressed &&
    priv->isHeld &&
    (priv->pressButton==0 || priv->pressButton==XFDASHBOARD_CLICK_ACTION_LEFT_BUTTON))
  {
    return(TRUE);
  }

  return(FALSE);
}
```

I haven't tested this piece of code and there might be syntax errors as well. Also all comments are missed. Maybe it works and helps.

gmc-holle commented:
Ah ok. Forget to test for `priv->isPressed` and `priv->isHeld` as we are only interested in to test for the right(tm) buttons or touch event. If it pressed was already checked before when the signal was emitted.

TheZoq2 commented:
I guess me doing that impementation wrong shows that a utility function would be a nice addition :)

I implemented that and now all the touch interractions seem to work as they should

gmc-holle commented on libxfdashboard/click-action.c:
```diff
@@ -887,3 +887,24 @@ void xfdashboard_click_action_release(XfdashboardClickAction *self)
+ * @self: A #XfdashboardClickAction
+ *
+ * Checks if the specified click action is either a left button press or a single touch 'tap'
+ */
```
Please document the return value also even if everybody may know ;)

```
 *
 * Return value: Returns %TRUE if click action event is a left button press or a single touch 'tap', otherwise %FALSE.
```

TheZoq2 commented:
> Except for a minor documentation issue, everything looks very good to me and can be merged :)
> 
> It just one question: Is xfdashboard_click_action_is_left_button_or_touch() a good name for this handy utility function or would xfdashboard_click_action_is_left_button_or_tap() a better one?

I agree, tap is a better name than touch. I changed that and added the return value to the documentation.

> Does these patches also address the scrolling issue you have mentioned in your initial request?

They do not, and I suspect that it would be a more difficult change to make. Since scrolling requires draging, more events need to be handled and in this case, I don't think just doing the same thing as a mouse interaction makes sense. 

Second, you can already drag things in the lists in order to add programs to the qucik launch panel. Somehow a distinction would need to be made between draging to scroll and draging to "move icons".

gmc-holle commented:
>> Except for a minor documentation issue, everything looks very good to me and can be merged :)
>>
>> It just one question: Is xfdashboard_click_action_is_left_button_or_touch() a good name for this handy utility function or would xfdashboard_click_action_is_left_button_or_tap() a better one?
>
> I agree, tap is a better name than touch. I changed that and added the return value to the documentation.

Thanks. :)

>> Does these patches also address the scrolling issue you have mentioned in your initial request?
> 
> They do not, and I suspect that it would be a more difficult change to make. Since scrolling requires draging, more events need to be handled and in this case, I don't think just doing the same thing as a mouse interaction makes sense.
>
> Second, you can already drag things in the lists in order to add programs to the qucik launch panel. Somehow a distinction would need to be made between draging to scroll and draging to "move icons".

Sounds difficult and as I do not own any touchscreen device (yeah except the smartphone) I cannot implement or test anything touch-related. Looking at the API documentation of Clutter it may be possible to replace XfdashboardClickAction with ClutterTapAction (or a new XfdashboardTapAction class to be still backwards compatible with Clutter 1.12) and add ClutterPanAction to XfdashboardViewpad to support "pan" gesture action for scrolling. But as said "maybe".

I hope the scrollbars work on your touchscreen device.

So far I will merge your patch set and I have to thank you for your work :) Thanks!

TheZoq2 commented:
No problem. I might have a look at trying to implement scrolling as well at some point. The scrollbars don't seem to work either, but the application lists aren't a huge feature for me. I mostly use the panel or search :)

gmc-holle commented:
Just one further question ;)

You referenced issue #156 . Is it resolved with your work?

TheZoq2 commented:
Nope, it is not. The reason I mentioned it was that I thought maybe the prints that I saw was because of these changes.
